src/pay/batch_payer.py

```diff
@@ -336,9 +347,13 @@ def simulate_single_operation(self, payment_item, pymnt_amnt, branch, chain_id):
             logger.debug("Error while validating operation - Status: {}, Message: {}".format(status, op_error))
             return PaymentStatus.FAIL, []
 
+        consumed_gas += GAS_SAFETY  # Tezos client does this
+        consumed_storage += STORAGE_SAFETY
+
         # Calculate needed fee for extra consumed gas
-        tx_fee += int(consumed_gas * MUTEZ_PER_GAS_UNIT)
-        simulation_results = consumed_gas, tx_fee, storage
+        tx_fee += int((consumed_gas - int(self.gas_limit)) * MUTEZ_PER_GAS_UNIT)
```

Comment:
I think we need to increase the fee also with some fee offset, many transactions are not being included currently

Reply:
Should I remove the `-self.gas_limit` then? That would add a bit back to the fee.

Reply:
In the short term, this will work. I highly suspect that playing whack-a-mole with smart contracts isn't going to be efficient long term, and I think TRD would benefit from moving to an approach where estimates are done dynamically.  

There's a few things you can do to make this work. Any of the following is probably worth doing in isolation, but taken together they make up a pretty decent solution

- It's easy to estimate a gas and storage limit for an operation using a `pre-apply` (if you need help to get this working, LMK)
- Once you know fees / gas for each operation, you can batch them up according two two measures (i) max operations in a batch [1] and (ii) gas limit

```
const batches: Array<Array<operation>> = []
let currentBatch = 0
let currentGasUsedInBatch = 0
for (i = 0; i < operations.length; i++) {
  const operation = operations[i]

  // Init a new batch if needed
  if (batches[currentBatch] == undefined) {
    batches[currentBatch] = []
  }
  
  // Start a new batch if we've hit the max in a batch
  if (batches[i].length > OPS_PER_BATCH) {
    currentBatch += 1
    currentGasUsedInBatch = 0
  }
  
  // Start a new batch if we hit the gas limit
  if (currentGasUsedInBatch + operation.gas_limit > OP_GAS_LIMIT) {
    currentBatch += 1
    currentGasUsedInBatch = 0
  }
    
  batches[currentBatch].push(operation)
  currentGasUsedInBatch += operation.gas_limit 
}

```
(please excuse the fact that I don't speak python)

- Once you have batches of ops, you can estimate fees offline using static analysis. Here's an example implementation in JS: https://github.com/tacoinfra/harbinger-lib/blob/master/src/operation-fee-estimator.ts. You could optionally configure this to take a price per gas unit, which is probably a better measure than an offset since that is the metric bakers are going to look at. 


[1] In the short term, it probably makes sense to make this a configurable command line option, since I think other bakers may soon hit this problem and it's the easiest fix.

Reply:
thank you for the comment keefer! the first step is already being performed correctly (gas+storage estimation), any failure related to gas/storage is detected and reported already at run_operation stage. Since problems are happening at the inclusion stage, the problem is most probably related to fee adjustment. 
![image](https://user-images.githubusercontent.com/33556352/115158034-86a8b100-a08c-11eb-9118-6b25ef1c650b.png)
Here you keep iterating till current fee is lower than required fee, I do not see however what is the changing parameter between two different iterations (head block?)